### PR TITLE
Use `storage::append` in the implementation of the storage types

### PR DESCRIPTION
**Pull request by bkchr**
This pr moves all the `append_*` functions to use the new `storage::append` functionality. This should improve the performance of all these calls. Besides that the host implementation is changed to use the `EncodeAppend` from `parity-scale-codec` instead of copying it.

The implementation comes with some "disadvantages":
- If the storage item is invalid, `append` will overwrite it and also ignores any default value set in the runtime. Sadly there isn't a better solution for this, but we also should teach people that vectors etc should not get any special default values.
- It now can only `append` one item at a time. Up to now, there was only one call to `append` that tried to append more items.

### CODE REVIEW

File: frame/support/test/tests/decl_storage.rs [hunk 1]
```diff
@@ -526,47 +526,47 @@ mod test_append_and_len {
 	#[test]
 	fn append_works() {
```
Comment (NikVolf) [inline frame/support/test/tests/decl_storage.rs:527]: kill + append
overwrite + append

would be nice additions to this test

Comment (bkchr) [reply frame/support/test/tests/decl_storage.rs:527]: > overwrite + append
Is done below in `append_overwrites_invalid_data`

Comment (bkchr) [reply frame/support/test/tests/decl_storage.rs:527]: Added kill + append

**Comment by NikVolf** [comment]
bench import/full-wasm

**Comment by gavofyork** [comment]
polkadot not building?

**Comment by bkchr** [comment]
Yeah, already working on a companion.

**Pull request merged**